Add documentation for periodic metrics files removal (#8552)
nautobot/nautobot: nautobot/docs/user-guide/administration/guides/prometheus-metrics.md

When deploying Nautobot in a multi-process manner (e.g. running multiple uWSGI workers) the Prometheus client library requires the use of a shared directory to collect metrics from all worker processes. To configure this, first create or designate a local directory to which the worker processes have read and write access, and then configure your WSGI service (e.g. uWSGI) to define this path as the `prometheus_multiproc_dir` environment variable.

> The downside of this is that you will have to ensure that this directory is cleaned up on a regular basis, as the Prometheus client library will create a separate file for each metric in each worker process, and these files are not automatically removed which in turn can lead to degradation of performance over time. You can use a cron job or similar scheduled task to periodically clean up this directory, for example:

Reviewer: "and these files are not automatically removed which in turn can lead to degradation of performance over time."

We should clarify that this happens when:
- There is a worker process churn (processes are re-created with a new PID).
- The processes are long-lived (e.g., k8s pod is not regularly restarted)

Also, might be worth mentioning that setting a sensible number of workers can help with reducing the number of files.

Reviewer: In that case, I think the following might not be factually correct. If processes never restart, we won't have a need for removing files. We might need to tighten the language here.

" and these files are not automatically removed which in turn can lead to degradation of performance over time"

> The downside of this is that you will have to ensure that this directory is cleaned up on a regular basis, as the Prometheus client library will create a separate file for each metric in each worker process, and these files are not automatically removed which in turn can lead to degradation of performance over time. You can use a cron job or similar scheduled task to periodically clean up this directory, for example:

Author: It is metric type indeed, now changed.

1. Create a Python script that scans the multiproc directory and removes files belonging to PIDs that are no longer running.

   ```python
   import os
   import re
   import shutil
   from prometheus_client import multiprocess

>    def cleanup_prometheus_orphans(metrics_dir):

Reviewer: ```suggestion
   def cleanup_orphaned_prom_metric_files(metrics_dir):
```

       """
       Scans the multiproc directory and removes files
       belonging to PIDs that are no longer running.
       """
       if not os.path.exists(metrics_dir):
           return

       # Pattern to find PIDs in filenames (e.g., gauge_multiproc_123.db)
       pid_pattern = re.compile(r'.+_(\d+)\.db$')

       # Get list of currently running PIDs
       active_pids = set()
       for pid in os.listdir('/proc'):
           if pid.isdigit():

>                active_pids.add(int(pid))

Reviewer: Let's add `try..except` for that `int`

Reviewer: `isdigit` operates on Unicode (see https://docs.python.org/3/library/stdtypes.html#str.isdigit), so unlikely, but possible to have `int()` raise an exception here.

When deploying Nautobot in a multi-process manner (e.g. running multiple uWSGI workers) the Prometheus client library requires the use of a shared directory to collect metrics from all worker processes. To configure this, first create or designate a local directory to which the worker processes have read and write access, and then configure your WSGI service (e.g. uWSGI) to define this path as the `prometheus_multiproc_dir` environment variable.

> Since the files stored in the designated directory are not meant to be long-lived, it is recommended to use a temporary directory such as `/tmp/nautobot_prometheus` or an `emptyDir` in Kubernetes environments for this purpose. This directory must be wiped between process runs (before startup is recommended). Another approach is to use a "script" to do it through a uwsgi hook before Workers accept requests. As an example:

Reviewer: ```suggestion
Since the files stored in the designated directory are not meant to be long-lived, it is recommended to use a temporary directory such as `/tmp/nautobot_prometheus` or an `emptyDir` in Kubernetes environments for this purpose. This directory must be wiped between process runs (before startup is recommended). Another approach is to use a script executed via a uWSGI hook before the first uWSGI worker starts accepting requests. As an example:
```

```ini
; uwsgi.ini
; Before first worker starts accept request
hook-accepting1 = exec:bash -c 'if [[ $prometheus_multiproc_dir ]]; then rm $prometheus_multiproc_dir/*.db; else echo "No prometheus multi_proc_dir"; fi'
```

> The downside of this is that you will have to ensure that this directory is cleaned up on a regular basis, as the Prometheus client library will create a separate file for each metric type within each worker process, and these files are not automatically removed which in turn, under certain conditions, may lead to degradation of performance over time. You can use a cron job or similar scheduled task to periodically clean up this directory, for example:

Reviewer: ```suggestion
The downside of this is that you will have to ensure that this directory is cleaned up regularly, as the Prometheus client library will create a separate file for each metric type within each worker process. These files are not automatically removed, which, in turn, under certain conditions, may lead to degradation of the performance of the metrics endpoint over time. You can use a cron job or similar scheduled task to periodically clean up this directory, for example:
```

> The downside of this is that you will have to ensure that this directory is cleaned up on a regular basis, as the Prometheus client library will create a separate file for each metric type within each worker process, and these files are not automatically removed. In turn, under certain conditions, this may lead to significant degradation of performance over time. For these situations where it's not enough to rely on cleanups based on periodic worker restarts, a more fitting approach is to clean up in a "live" manner. You can use a cron job or similar scheduled task to periodically clean up this directory, for example:

Author: Rephrased it, please check.
